### ISSUE TITLE
Make a PRE_RUN_ALLOW_FAIL option

### ISSUE THREAD

**Issue by bemsha** [Nextdoor/kingpin#244]
Kingpin's default behavior is to perform a dry run before each non-dry run. If the dry run fails at a point in the script after desirable work would have been done, it still aborts the run without doing any work.

I'd like to be able to override this abort.

**Comment by siminm** [comment]
I think this calls for a discussion of why you'd be ok with a failed dry run. We already have Conditions ( #70 ) and Warn on Failure ( #111 ) features. We've been under the impression that those two suffice any flexibility you want to achieve.

**Comment by diranged** [comment]
I've wavered back and forth on this issue ... this is actually the 3rd time I've written this response! :). Here's my $0.02...

@siminm, I agree with you that `warn_on_failure` can help here ... but its not perfect. If @bemsha sets `warn_on_failure: %IGNORE_FAILURE%` in his outer-most *group.Sync* actor, then indeed a failure during the *dry* run will fail gracefully and allow the script to begin its *real* run. However, because the same setting is used in the *real* run, a failure in this real run will also be warned rather than errored on.

@bemsha, I believe that we could add a global `DISABLE_DRY_RUN` environment variable that completely prevents the initial dry run. If you're going to ignore the dry run output/sanity check anyways, then executing the run makes little sense in the first place. This setting can be implemented easily, and should be **rarely** used.

@bemsha, one quick question though. Historically we have avoided this type of option because its typically an indication that the actors themselves are not flexible enough. Can you tell us the specific use case where you blocked and need this functionality so that we can try to address the actor issues themselves?

**Comment by bemsha** [comment]
Conditions require adding a lot of variables to the script and to the Jenkins driver to give control at sufficiently fine grain to stop at any chosen place in the script.

I don't want the script to warn and continue when something goes wrong. I want it to stop right there and report failure. It's not desirable to make it possible to perform a late step for which a prerequisite has not been completed.

**Comment by bemsha** [comment]
@diranged Last night I ran a Kingpin job that was meant to launch server arrays, add them to ELBs, mark the new arrays as the "live" service, and tear down the old arrays.

The dry run told me that the job would fail in the teardown step. To perform the steps up to this failure point I edited the JSON file that configures this job.

As it turned out, the run with the edited JSON file failed at an earlier step than the dry run had predicted. This failure was not at a stage boundary, so simple Conditions wouldn't have given the desired result.  Warn-on-failure would have allowed later steps to run that shouldn't have run.

The warn-on-failure mechanism would be more useful if there were a way to specify that one action is a hard prerequisite for another action, and that the second action must not be performed unless the first action succeeds. I don't think we want to add this much complexity.
